route-rs-packets/src/arp.rs

```diff
@@ -0,0 +1,253 @@
+use crate::{EthernetFrame, MacAddr};
+use std::convert::TryInto;
+use std::net::IpAddr;
+
+pub enum ArpOp {
+    Request = 1,
+    Reply = 2,
+}
+
+pub enum ArpHardwareType {
+    Ethernet = 1,
+}
+
+// TODO: should move to types.rs or ethernet.rs
+pub const IPV4_ETHER_TYPE: u16 = 0x0800;
+pub const IPV6_ETHER_TYPE: u16 = 0x86DD;
+pub const ARP_ETHER_TYPE: u16 = 0x0806;
+
+const HARDWARE_TYPE_RANGE: (usize, usize) = (0, 2);
+const PROTOCOL_TYPE_RANGE: (usize, usize) = (2, 4);
+const HARDWARE_ADDR_LEN_RANGE: (usize, usize) = (4, 5);
+const PROTOCOL_ADDR_LEN_RANGE: (usize, usize) = (5, 6);
+const OPCODE_RANGE: (usize, usize) = (6, 8);
+
+///
+/// EthernetFrame wrapper with getters/setters for the packet structure described in RFC 826
+/// https://tools.ietf.org/html/rfc826
+///
+// NOTE: Could be implemented in various ways, such as a specialized version of EthernetFrame that's
+// known to be an ARP frame. It could be implemented in a similar way that packets are promoted/demoted
+// with `TryFrom`.
+#[derive(Clone)]
+pub struct ArpFrame {
+    frame: EthernetFrame,
+}
+
+impl ArpFrame {
+    // TODO: would like default args for Ethernet (6 bytes) and IPv4 (4 bytes)
+    // TODO: Should we accept a struct of options to initialize the ArpFrame with? Or use a builder pattern?
+    pub fn new(hardware_addr_len: u8, protocol_addr_len: u8) -> Self {
+        let payload_len = 8 + (2 * hardware_addr_len as usize) + (2 * protocol_addr_len as usize);
+        let payload: Vec<u8> = vec![0; payload_len];
+
+        let mut frame = EthernetFrame::empty();
+        frame.set_payload(payload.as_slice());
+
+        let mut arp_frame = ArpFrame { frame };
+        arp_frame.set_hardware_addr_len(hardware_addr_len);
+        arp_frame.set_protocol_addr_len(protocol_addr_len);
+        arp_frame
+    }
+
+    pub fn from(frame: EthernetFrame) -> Self {
```

Comment:
Let's not reimplement from, but not in the From trait, that is confusing.  Also I suggest TryFrom instead of From, since not all EthernetFrame's are ArpFrame's and we should reflect that in our interface.

Reply:
Good idea, I didn't think of that. Should we support both

`try_from(frame: EthernetFrame)` and `from_buffer(buf: PacketData)`?

Reply:
Yeah, typically, I've just made `try_from()` call `from_buffer()` and left all the validation logic inside of `from_buffer`.  If you decide to keep wrapping an EthernetFrame rather than have the raw vector, you may need something slightly different.